### PR TITLE
[clang] Add/enhance documentation for some important classes.

### PR DESCRIPTION
**Pull request by sam-mccall**

### CODE REVIEW

**Comment by sam-mccall** [comment]
There are some classes here where I think the concepts are tricky and under-explained (SourceManager, TypeLoc) or just important enough that it's worth spending a few more words in general.
(I'm sure there are a bunch more, but I hit my timebox)

If you think anything is too wordy/confusing or I'm removing anything important, please do say so!

Let me know if it's better to split these into smaller patches.

File: clang/include/clang/Basic/SourceManager.h [hunk 1]
```diff
@@ -648,18 +648,54 @@ class InBeforeInTUCacheEntry {
...
+/// A SourceLocation can point at any byte of code. Rather than store
+/// (file, offset) pairs, it is a single offset into a buffer of all files
+/// concatenated together.
+///
+/// [--file1--][----file2----][file3]
+///               ^
+/// This buffer does not exist in memory. Instead SourceManager keeps an array
+/// of SLocEntry objects, each describing one file and its offset in the buffer.
+/// Each entry is assigned a FileID, and SourceManager can encode/decode a
+/// SourceLocation into a (FileID, file offset) pair: see getDeomposedLoc().
```
Comment (aganea) [inline clang/include/clang/Basic/SourceManager.h:680]: typo: getDecomposedLoc() perhaps?

File: clang/include/clang/Basic/SourceManager.h [hunk 2]
```diff
@@ -648,18 +648,54 @@ class InBeforeInTUCacheEntry {
...
+///
+/// The SourceManager sits above the FileManager, which reads source files and
+/// exposes them as FileEntrys.
+/// SourceManager does not generally know about tokens or AST nodes, the lexer
+/// and parser are layered above SourceManager.
+/// 
+/// ====== SourceLocations, FileIDs, and SLocEntrys =======
+///
+/// A SourceLocation can point at any byte of code. Rather than store
+/// (file, offset) pairs, it is a single offset into a buffer of all files
```
Comment (aganea) [inline clang/include/clang/Basic/SourceManager.h:672]: it is a single offset into a **virtual** buffer of all files?

File: llvm/include/llvm/Support/VirtualFileSystem.h [hunk 3]
```diff
@@ -7,7 +7,10 @@
 //===----------------------------------------------------------------------===//
 //
 /// \file
-/// Defines the virtual file system interface vfs::FileSystem.
+/// Provides abstract filesystem APIs to decouple from OS-level file access.
+///
+/// This allows replacing access to physical files on disk by in-memory data,
+/// for testing or other purposes.
```
Comment (aganea) [inline llvm/include/llvm/Support/VirtualFileSystem.h:13]: for testing, **caching** or other purposes.

File: clang/include/clang/AST/Stmt.h [hunk 4]
```diff
@@ -79,8 +79,24 @@ enum class StringLiteralKind;
...
+///
+/// Statements can have children and so form trees. e.g. `while (i>0) i--;`
+///
+///   WhileStmt
+///   |-BinaryOperator >
+///   | |-DeclRefExpr i
+///   | `-IntegerLiteral 0
+///   `-UnaryOperator --
+///       DeclRefExpr i
+///
```
Comment (zwuis) [inline clang/include/clang/AST/Stmt.h:97]: We can use `\code` and `\endcode` commands for code block so that it is treated as code instead of text by doxygen.

```cpp
/// \code
///   WhileStmt
///   |-BinaryOperator >
///   ...
/// \endcode
```

Comment (sam-mccall) [reply clang/include/clang/AST/Stmt.h]: Thanks, this was indeed being mangled.

I've instead indented by two more spaces. Doxygen understands this part of markdown, and I think it keeps the source file more readable.

**Review by kadircet — APPROVED** [review]
thanks a lot for doing this!
